### PR TITLE
Introduce TMCReplay and restructure repo

### PR DESCRIPTION
**Pull request by benedikt-voelkel**
[TMCReplay]

Simulation engine based on the Virtual Monte Carlo (VMC) framework to
exactly replay the steps as previously recorded from another run using
the MCStepLogger.

Cut and process settings can be varied to study the impact on the steps.
That funcionality is currently not fully implemented but will be as the
next step.

The replay engine is built by default but the build can be switched off
setting `-MCStepLogger_BUILD_TMCReplay=OFF" at configure time when
`CMake` is invoked.

In the easiest case it can be run using the installed executable
`tmcreplay`. Use `tmcreplay --help` to see all options.

Generally, the engine can be used as usual within any other simulation
framework. In that case it might be used to evaluate the impact of
different cut and process settings on hist or digits.

[MCStepLogger]

3-momenta (px, py, pz) added to logging as that is necessary to replay
simulation correctly

[Repository restructuring]

The original MCStepLogger (and the TMCReplay engine) were moved to
corresponding sub-directories.

Inistallation of libraries was unified through common CMake function.

Apply clang-format to cxx (previously not done).

### CODE REVIEW

File: TMCReplay/src/replay.cxx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+#include "TMCReplay/GenericApplication.h"
+#include "TMCReplay/GenericStack.h"
+#include "TMCReplay/Engine.h"
+
+namespace bpo = boost::program_options;
+
+int main(int argc, char* argv[])
+{
+  // prepare reading cmd options and store them
+  bpo::options_description desc("Simulation options");
```
Comment (sawenzel) [inline TMCReplay/src/replay.cxx]: I would add a longer sentence of what this executable is doing. "Replaying a previously recorded step by step. Producing the same output as reference simulation. Meant for checking / performance measurements ..."

File: TMCReplay/src/replay.cxx [hunk 2]
```diff
@@ -0,0 +1,54 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
```
Comment (sawenzel) [inline TMCReplay/src/replay.cxx]: The filename is lower case here whereas others start with a capital letter. Better to be consistent.

**Comment by sawenzel** [comment]
Looks very good and it compiles. 2 generic comments
a) Why "TMCReplay" ? The "T" confuses me. Maybe we could call the C++ class "TVMCReplay" just to be consistent with "TGeant4" .. as an implementation of VMCEngine but on the other hand this is not set in stone. I would definitely leave out the "T" in executables etc.

b) Can we have the test data be public on gitlab, not requiring authentication.

File: TMCReplay/README.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+### Running in another environment
+
+In that case the geometry construction is assumed to implemented by that environment's implementation of `TVirtualMCApplication`. Also, no stack is constructed automatically and it has to be constructed and passed by the user.
+
+Such a scenario is useful to evaluate the impact of parameter variation on hits or even digits.
+
+
+## Comparing 2 step files after a replay
+
+If you want to compare the original step file with the one derived from the replay go [here](../README.md#comparing-analysis-output).
```
Comment (sawenzel) [inline TMCReplay/README.md]: does this link exist?

**Comment by benedikt-voelkel** [comment]
Thanks for the comments @sawenzel 
I also managed in the meantime to link O2 against the replay engine. It also runs but complains about a few methods (`::Material`, `::Mixture` and `::Medium`) not yet implemented which I am taking care of right now.
I propose after addressing your current set of comments and implemented the missing functionality, we merge this since it is self-consistent as it is. Also, since we are using tags now in `alidist` it is safe to do so.
After that
1. we can tackle the actual implementation of the energy cuts. What do you think?
1. test the actual usage with O2 and add required functionality

File: TMCReplay/src/replay.cxx (hunk 2 above)
Comment (benedikt-voelkel) [reply TMCReplay/src/replay.cxx]: That's the one compiled into an executable, following what we had before for the macros and executables in the `MCStepLogger `. Should they all be capitalised?

File: TMCReplay/README.md (hunk 3 above)
Comment (benedikt-voelkel) [reply TMCReplay/README.md]: No and removed

**Comment by benedikt-voelkel** [comment]
Hej @sawenzel 

Concerning the naming convention from **a)** above
How about
1. everything is put into the namespace `mcreplay`and in there have
    1. `mcreplay::Engine`
    1. `mcreplay::GenericApplication`
    1. `mcreplay::GenericStack`
1. the top-level directory could the be called `MCReplay` cause that would then be aligned with `MCStepLogger` where everything is inside the namespace `mcsteplogger` (indeed, everything there is **actually** in the namespace `o2::mcsteplogger`, we might want to change that and remove the `o2` in the near future?!)

concerning **b)**: The test data repo is now public

**Comment by sawenzel** [comment]
Concerning the namespace: I am not a big fan because "::Engine" is not informative enough  (if somebody is qualifying the class without namespace). Also "GenericApplication / GenericStack" sound too generic (why do we need them by the way?). 
I'd prefer to call the classes what they are: For instance "VMCReplayer[Engine]". I think we should make a reference to "VMC" ... so in the end not saying "TMC" but "VMC". We can still have the project namespace in addition.

Yes, we can get rid of `o2::`

**Comment by benedikt-voelkel** [comment]
The `GenericStack` and `GenericApplication` are there in order to run the replay engine fully stand-alone from any other environment. Just imagine one is really only interested in steps, then having a lightweight application and stack at hand I thought would be nice.
This is exactly what one could do with the executable built from `src/replay.cxx`

Namespace and other naming: I agree, when using the namespace, `Engine` is pretty much saying nothing. So I will follow your suggestion.

File: TMCReplay/src/replay.cxx (hunk 2 above)
Comment (benedikt-voelkel) [reply TMCReplay/src/replay.cxx]: @sawenzel can this be marked as resolved or should we rename the executable sources to also start with caps?

**Comment by benedikt-voelkel** [comment]
The naming convention has been revised
* `tmcreplay::Engine` -> `mcreplay::MCReplayEngine`
* `tmcreplay::GenericApplication` -> `mcreplay::MCReplayGenericApplication`
* `tmcreplay::GenericStack` -> `mcreplay::MCReplayGenericStack`

Another class `mcreplay::MCReplayEvGen` has been added.

The stack, application and event generator were added to make it possible to run a stand-alone replay based on the steps only.

For now, the geometry construction (including materials and media) follows https://github.com/vmc-project/vmc/blob/master/source/include/TGeoMCGeometry.h

In addition various revisions and fixes were implemented to guarantee the correct replay

**Pull request merged**